src/gui/toolbarMenubar/AbstractSliderItem.cpp

```diff
@@ -0,0 +1,194 @@
+    publicApi_->configure(slider, horizontal);
+
+    gtk_container_add(GTK_CONTAINER(it), GTK_WIDGET(slider));
+    return {it, slider};
+}
+
+void AbstractSliderItem::Impl::setSlider(GtkRange* slider) {
+    GtkWidget* newSliderWidget = GTK_WIDGET(slider);
+
+    // If we already had a slider, disconnect from its events.
+    if (slider_) {
+        GtkWidget* oldSliderWidget = GTK_WIDGET(slider_);
+        g_signal_handlers_disconnect_by_func(oldSliderWidget, (void*)(onSliderChanged), publicApi_);
+        g_signal_handlers_disconnect_by_func(oldSliderWidget, (void*)(onSliderButtonPress), publicApi_);
+        g_signal_handlers_disconnect_by_func(oldSliderWidget, (void*)(onSliderButtonRelease), publicApi_);
+        g_signal_handlers_disconnect_by_func(oldSliderWidget, (void*)(onSliderHoverScroll), publicApi_);
+        g_signal_handlers_disconnect_by_func(oldSliderWidget, (void*)(formatSliderValue), publicApi_);
+    }
+
+    // Connect new events.
+    g_signal_connect(newSliderWidget, "value-changed", G_CALLBACK(onSliderChanged), publicApi_);
+    g_signal_connect(newSliderWidget, "button-press-event", G_CALLBACK(onSliderButtonPress), publicApi_);
+    g_signal_connect(newSliderWidget, "button-release-event", G_CALLBACK(onSliderButtonRelease), publicApi_);
+    g_signal_connect(newSliderWidget, "scroll-event", G_CALLBACK(onSliderHoverScroll), publicApi_);
+    g_signal_connect(newSliderWidget, "format-value", G_CALLBACK(formatSliderValue), publicApi_);
+
+    slider_ = slider;
+}
+
+void AbstractSliderItem::Impl::setEnabled(bool enabled) {
+    if (slider_) {
+        gtk_widget_set_sensitive(GTK_WIDGET(slider_), enabled);
+    }
+    if (toolItem_) {
+        gtk_widget_set_sensitive(GTK_WIDGET(toolItem_), enabled);
+    }
+}
+
+void AbstractSliderItem::Impl::onSliderChanged(GtkRange* range, AbstractSliderItem* parent) {
+    parent->onSliderChanged(parent->scaleFuncInv(gtk_range_get_value(range)));
+}
+
+bool AbstractSliderItem::Impl::onSliderButtonPress(GtkRange* range, GdkEvent* event, AbstractSliderItem* parent) {
+    parent->onSliderButtonPress();
+
+    return false;  // Let the slider handle the event.
+}
+
+bool AbstractSliderItem::Impl::onSliderButtonRelease(GtkRange* range, GdkEvent* event, AbstractSliderItem* parent) {
+    parent->onSliderButtonRelease();
+    return false;
+}
+
+bool AbstractSliderItem::Impl::onSliderHoverScroll(GtkRange* range, GdkEventScroll* event, AbstractSliderItem* parent) {
+    parent->onSliderHoverScroll();
+    return false;
+}
+
+auto AbstractSliderItem::Impl::formatSliderValue(GtkRange* range, double value, AbstractSliderItem* parent) -> gchar* {
+    std::string text = parent->formatSliderValue(parent->scaleFuncInv(value));
+
+    // TODO: At present, gchar* is equivalent to (const char*). Is there a better/won't break
+    // way of doing this cast?
+    return g_strdup(static_cast<const gchar*>(text.c_str()));
```

Comment:
I don't like this conversion from `std::string` to `gchar*`. Is there a better way to do this?